api/v1beta1/preflightvalidation_types.go

```diff
@@ -100,6 +71,58 @@ type PreflightValidation struct {
 	Status PreflightValidationStatus `json:"status,omitempty"`
 }
 
+func (p *PreflightValidation) ConvertTo(dstRaw conversion.Hub) error {
+	dst := dstRaw.(*v1beta2.PreflightValidation)
+
+	dst.ObjectMeta = p.ObjectMeta
+	dst.Spec = p.Spec
+
+	dst.Status = v1beta2.PreflightValidationStatus{}
+
+	if count := len(p.Status.CRStatuses); count > 0 {
+		dst.Status.Modules = make([]v1beta2.PreflightValidationModuleStatus, 0, count)
+
+		for k, v := range p.Status.CRStatuses {
+			namespace, name, ok := strings.Cut(k, "/")
+
+			if !ok || v == nil {
+				// Elements whose key is not a namespace name or that are nil are invalid.
+				return nil
+			}
+
+			status := v1beta2.PreflightValidationModuleStatus{
+				CRBaseStatus: *v,
+				Namespace:    namespace,
+				Name:         name,
+			}
+
+			dst.Status.Modules = append(dst.Status.Modules, status)
+		}
+	}
+
+	return nil
+}
+
+func (p *PreflightValidation) ConvertFrom(srcRaw conversion.Hub) error {
+	src := srcRaw.(*v1beta2.PreflightValidation)
+
+	p.ObjectMeta = src.ObjectMeta
+	p.Spec = src.Spec
+
+	p.Status = PreflightValidationStatus{}
+
+	if count := len(src.Status.Modules); count > 0 {
+		p.Status.CRStatuses = make(map[string]*CRStatus, count)
+
+		for _, v := range src.Status.Modules {
+			v := v
```

Comment:
Why are we shadowing `v` here with a new variable named `v1`?

Reply:
Not sure I get the question, there is no `v1` here.
We need to do `v := v` because the loop variable is a pointer. If we did not have that line, all elements in the map would point to the same struct. Go [changes this behaviour in 1.22](https://go.dev/blog/loopvar-preview), but in the meantime we still need to do this.

Reply:
I don't think it would. I think that you are referring to [closures](https://gobyexample.com/closures) that are evaluated, apparently, only when called.

In this case, this is a map (not a function/goroutine capturing a closure), therefore, I believe it will work without `v := v` (which is very confusing).

See [this playground](https://go.dev/play/p/S_AGdmKkPpN?v=goprev)

Reply:
I don't think the playground reproduces what we are doing here. [That one](https://go.dev/play/p/H881vVkZZGX?v=goprev) exhibits the problem.

Reply:
Ohh OK, I see what the issue is, it is different than the issue I remembered. Here `v` in the loop is the same variable through the whole loop so each iteration actually override all prior iterations since it is using the same variable.

I remember that the "go to" fix was indeed to do `v := v` - I am very happy they fixed it in 1.22!